File: code/1227/1227分巧克力.py
h, w = [], []


def check(x):
    cnt = 0
    for i in range(n):
        cnt += (h[i] // x) * (w[i] // x)
        if cnt >= k:
            return True

    return False


if __name__ == "__main__":
    n, k = map(int, input().split())
    for i in range(n):
        a, b = map(int, input().split())
        h.append(a)
        w.append(b)

    l, r = 1, int(1e5)
    while l < r:
        mid = l + r + 1 >> 1
        if check(mid):
            l = mid
        else:
            r = mid - 1

    print(l)

File: code/1227/1227分巧克力.cpp
#include <algorithm>
#include <cstring>
#include <iostream>
using namespace std;

const int N = 1e5 + 10;
int h[N], w[N];
int n, k;

bool check(int x) {
  int cnt = 0;
  for (int i = 0; i < n; i++) {
    cnt += (h[i] / x) * (w[i] / x);
    if (cnt >= k)
      return true;
  }

  return false;
}

int main() {
  cin >> n >> k;
  for (int i = 0; i < n; i++)
    cin >> h[i] >> w[i];

  int l = 1, r = 1e5;
  while (l < r) {
    int mid = l + r + 1 >> 1;
    if (check(mid))
      l = mid;
    else
      r = mid - 1;
  }

  cout << l << endl;

  return 0;
}
